[5.9][move-only] A set of batched changes

gottesmm opened the pull request.

gottesmm committed:
[move-only] If we diagnose an invalid escaping capture, turn off noncopyable checking in the closure.

The reason why I am doing this is that currently SILGen knows when emitting said
closure that we are going to emit an error (that it does not have enough
information to emit itself since it doesn't know the caller), so doesn't emit
move checking markers. The result is that the move checker will not eliminate
any copies in the closure and thus will emit a "copy of noncopyable type found"
error and tell the user to file a bug. This just suppresses that.

rdar://108511866
(cherry picked from commit b585a1870fecad5d8aaa63c92839c1826b7a1514)

gottesmm committed:
[move-only] When emitting borrows for move only types, use a load [copy] instead of a load_borrow.

The reason why I am doing this is that otherwise if one has a function that
takes both a guaranteed and an owned parameter, we will break OSSA invariants
since the load [take] will invalidate the load_borrow. So instead, we put in a
load_borrow knowing that the move checker will convert it to a load_borrow
assuming that the two pass exclusivity checking.

NOTE: Because of some missing functionality in subsequent tests, I had to
disable one test (moveonly_escaping_definite_initialization.swift) and also add
some checks for copy of noncopyable object errors. They will go away in the next
2 commits.

rdar://108510987
(cherry picked from commit 438974a2b98ec73b59a71625943710a931b0a0b4)

gottesmm committed:
[move-only] Instead of using AccessUseVisitor to visit addresses use TransitiveAddressVisitor.

This makes it so that the move address checker is not dependent on starting the
traversal at a base object. I also included verifier checks that the API can
visit all address uses for:

1. project_box.
2. alloc_stack.
3. ref_element_addr.
4. ref_tail_addr.
5. global_addr_inst.

this is because this visitor is now apart of the SIL API definition as being
able to enumerate /all/ addresses derived from a specific chosen address value.

This is a refactoring NFCI change.

rdar://108510644
(cherry picked from commit c577055100e26616fe68c7ad68600ce864b4f2c2)

gottesmm committed:
[generic-spec] Add an option -sil-generic-dump-functions-after-specialization that makes it easy to look for errors around specialization.

(cherry picked from commit b957fbb4df0bde6e82e46ba495137ddcd8e343c0)

gottesmm committed:
[move-only] Eliminate dead code that was not removed when re-projected move only types were removed.

(cherry picked from commit 9252d5ea8b9cfcb72fd069eb552c61cbd962d067)

gottesmm committed:
[move-only] When emitting accesses to let boxes containing a noncopyable type, always emit mark_must_check.

The reason to do this is that:

1. Otherwise, we do not emit markers when someone attempts to consume the let.
We need the no_consume_or_assign to be there.
2. We need to insert assign_but_not_consuming so that DI can properly check lets
that are conditionally initialized and convert them to
initable_but_not_consuming.

I included a full definite_init SIL test that validates that we get the correct
codegen after DI in this case and emit the appropriate error as well.

rdar://108511534
(cherry picked from commit 9a0b9665f1be2e1640bd48ec4c9ca93e2ca86dfa)

gottesmm committed:
[move-only] Add a large SILGen/-verify test that checks captures of escaping closure behavior.

NOTE: We consider consuming closures that are marked without escaping as
non-escaping. We shouldn't do this, but for now I just included them so that
their behavior is in some test and we at least don't crash.

(cherry picked from commit af8a097c83061ce3af0fd4e57bacf4424937438e)

gottesmm committed:
[move-only] Clone all __owned tests into consuming tests.

I left in the __owned tests of course.

rdar://108511703
(cherry picked from commit e3054a003c71c3d5658ecdd73162ca98d13cbead)

gottesmm committed:
[move-only] When emitting exclusivity diagnostics for move only types, do not suggest to the user to make a local copy.

It doesn't make sense to give this note since one can't make a copy of a
noncopyable type.

rdar://108511627
(cherry picked from commit 7e992cef6cf3454aee3a9a92269601533093c114)

gottesmm committed:
[move-only] Convert __shared to borrowing in move only tests.

These are the same semantically, just the mangling is slightly different. The
benefit of doing this is that we are actually testing what we expect our users
to do.

rdar://108511703
(cherry picked from commit 739417ff090c00dbb23832097c53d446b2b767de)

gottesmm committed:
Update test for changes in https://github.com/apple/swift/pull/65377

That PR was for rdar://108385761

(cherry picked from commit 9c8d22403958f9a3f42bf82a3cee5d43478769d7)

gottesmm commented:
I am just starting the testing. Going to do the CCC before asking for review from the branch managers.

gottesmm commented:
@swift-ci test

kavon reviewed (approved):
LGTM as this is a cherry-pick of https://github.com/apple/swift/pull/65353

gottesmm committed:
Remove reference to DropDeinitInst since it has not been cherry-picked yet.

gottesmm commented:
@swift-ci test

gottesmm commented:
I removed a reference to drop_deinit since it has not been cherry-picked yet.

kavon commented:
Here's a cherry-pick for `drop_deinit` that we ought to get in first: https://github.com/apple/swift/pull/65449

gottesmm commented:
Kavon and I spoke and he agrees that given the small amount of change to the PR needed (I just removed it from the walker), I am going to get this in now.

gottesmm commented:
CCC. This PR contains a few small commits that do some small
cleanups, so I did not include them in the CCC (example: adding
more tests/adding an option to make it easier to debug the
generic specializer). The CCC for the major changes are as
follows:

Commit 74146394bba8631eca3cf295ddbce9f9fc1eeb48:

• Description: Change the checker from using AccessScopeVisitor
  to using TransitiveAddressVisitor. This fixes the ability for
  the checker to handle lets used by escaping closures. The
  AccessScopeVisitor needed either reprojecting to occur on every
  access or to have a new scope on every access. Lets do not have
  a new scope everytime, so that did not
  work. TransitiveAddressVisitor just works transitively from any
  specific address, so it does not have this problem.
• Risk: Low risk. Only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/65353
• Reviewed By: @atrick @kavon
• Testing: I added verification that the address visitor always
  works and also added tests that showed that lets worked.
• Resolves: rdar://108510644

Commit ac28fa2901ed6b0daf48d44d1936f558c586b3ca:
• Description: When emitting borrows for move only types, use a
  load [copy] instead of a load_borrow to prevent a verifier
  crash due to breaking Ownership SSA invariants. The reason why
  I am doing this is that otherwise if one has a function that
  takes both a guaranteed and an owned parameter, we will break
  OSSA invariants (causing a crash) since the load [take] will
  invalidate the load_borrow. So instead, we put in a load_borrow
  knowing that the move checker will convert it to a load_borrow
  assuming that the two pass exclusivity checking.
• Risk: Low risk. Only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/65353
• Reviewed By: @kavon
• Testing: I added tests that made sure that functions that take
  a value twice once as borrowed and second as consuming work and
  do not hit verifier asserts.
• Resolves: rdar://108510987

Commit 7b7fd4df1d2b5a6802302ddda4def877c120fe36
• Description: If we diagnose an invalid escaping capture, turn
  off noncopyable checking in the closure.  The reason why I am
  doing this is that currently SILGen knows when emitting said
  closure that we are going to emit an error (that it does not
  have enough information to emit itself since it doesn't know
  the caller), so doesn't emit move checking markers. The result
  is that the move checker will not eliminate any copies in the
  closure and thus will emit a "copy of noncopyable type found"
  error and tell the user to file a bug. This just suppresses
  that.
• Risk: Low. Only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/65353
• Reviewed By: @kavon
• Testing: I added tests that show this behavior in action and
  made sure that we do not emit the inappropriate error.
• Resolves: rdar://108511866

Commit c19938433f257a8d5ab37fd905eafe83096a89bd:
• Description: When emitting accesses to let boxes containing a
  noncopyable type, always emit mark_must_check. This makes it so
  that we properly handle conditionally initialized noncopyable
  types.
• Risk: Low. Only affects noncopyable types.
• Original PR: https://github.com/apple/swift/pull/65353
• Reviewed By: @kavon
• Testing: Added tests that exercise the code.
• Resolves: rdar://108511534

Commit a0596d1974b9ac3a972dbf4f8ca848d594f1b9e6
• Description: When emitting exclusivity diagnostics for move
  only types, do not suggest to the user to make a local copy.
• Risk: Low. This only changes the exclusivity checker in the
  case when it sees a noncopyable type.
• Original PR: https://github.com/apple/swift/pull/65353
• Reviewed By: @kavon
• Testing: Updated tests that provide code coverage for the diagnostic for the changed diagnostic.
• Resolves: rdar://108511627

tbkka commented on lib/SIL/Utils/OwnershipUtils.cpp:
```diff
@@ -1024,163 +1024,6 @@ bool BorrowedValue::visitInteriorPointerOperandHelper(
-// FIXME: This does not yet assume complete lifetimes. Therefore, it currently
-// recursively looks through scoped uses, such as load_borrow. We should
-// separate the logic for lifetime completion from the logic that can assume
-// complete lifetimes.
```
Was this FIXME addressed?

gottesmm replied:
This FIXME is actually related to how this API is used by some older Ownership SSA APIs, not APIs that are just attempting to use it to visit addresses and their transitive users. My change refactors the visitor construct so that it can express both the expected behavior of those Ownership SSA APIs and also handle use cases where we just want a simple address walker (which the move checker needed). Since the actual change needed in question here was to expose this address walker functionality rather than fix issues in the Ownership SSA based APIs, it made sense to just ensure that I preserved the current behavior for those APIs.

I did this by provided 2 versions of findTransitiveUsesForAddress:

1. A new one that takes a visitor construct and just visits transitive addresses and their users.
2. Then I have a 2nd API that has the original function signature and that is called by the Ownership SSA APIs. The 2nd API calls the 1st API with a visitor struct that preserves the original behavior of the API (and what the Ownership SSA API users expect) by having its visitor struct when it visits a use that is a scoped instruction, append the scope ending instructions instead.

atrick replied:
Basically, the FIXME says that this code is over-complicated because we still haven't enabled complete lifetimes. There are many such places to be cleaned up later. it should just have been a TODO

tbkka commented on lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp:
```diff
@@ -562,18 +568,35 @@ static void diagnoseExclusivityViolation(const ConflictingAccess &Violation,
-                     Ctx.SourceMgr);
+    if (isMoveOnly) {
+        auto D = diagnose(Ctx, MainAccess.getAccessLoc().getSourceLoc(),
+                          diag::exclusivity_access_required_moveonly,
```
Duplicating this entire block seems unnecessary if the only difference is the diagnostic wording.

gottesmm replied:
I did this b/c the arguments to the diagnostics are different.

tbkka commented on test/Constraints/moveonly_constraints.swift:
```diff
@@ -42,7 +42,7 @@ enum Maybe<T> {
 }
 
-func takeConcrete(_ m: __shared MO) {}
+func takeConcrete(_ m: borrowing MO) {}
```
Should we keep tests of `__shared` as well?   I expected these to be duplicated, not replaced.

gottesmm replied:
The only difference in between shared and borrowing is the mangling. Otherwise, I would have duplicated the tests like I did with consuming.

tbkka commented on test/SILGen/moveonly.swift:
```diff
@@ -71,12 +73,12 @@ public func consumeVal(_ s: __owned NonTrivialStruct2) {}
-// CHECK: } // end sil function '$s8moveonly19useNonTrivialStructyyAA0cdE0VhF'
-public func useNonTrivialStruct(_ s: __shared NonTrivialStruct) {
+// CHECK: } // end sil function '$s8moveonly19useNonTrivialStructyyAA0cdE0VF'
+public func useNonTrivialStruct(_ s: borrowing NonTrivialStruct) {
```
This mangling change surprises me.  I can see why `__shared` and `borrowing` might mangle differently, but we're just losing a character here, which looks like we're not including `borrowing` in the mangling at all?

tbkka replied:
CC: @jckarter on the mangling question.  What do we expect to tell people who use `__shared` currently?   If the mangling is affected, then they can't replace that with `borrowing`.  Do we expect to support `__shared` indefinitely?

gottesmm replied:
@tbkka we should still ask @jckarter but since he is on vacation, I will give you my recollection. My recollection is that borrowing only affects the mangling if it changes the default convention. But my memory may be incorrect.

tbkka reviewed (commented):
A couple of small points I'd like to understand.
